packages/runtime/container-runtime/src/opLifecycle/opDecompressor.ts

```diff
@@ -76,9 +85,42 @@ export class OpDecompressor {
             const intoString = new TextDecoder().decode(decompressedMessage);
             const asObj = JSON.parse(intoString);
 
-            return { ...message, contents: asObj[0] };
+            return {
+                message: newMessage(message, asObj[0]),
+                state: "Processed",
+            };
         }
 
-        return message;
+        return {
+            message,
+            state: "Skipped",
+        };
     }
 }
+
+// We should not be mutating the input message nor its metadata
+const newMessage = (originalMessage: ISequencedDocumentMessage, contents: any): ISequencedDocumentMessage =>
+    stripCompressionMarkers({
+        ...originalMessage,
+        contents,
+        metadata: { ...originalMessage.metadata },
+    });
+
+// After compression, it is irrelevant to the other layers whether or not the
+// original message was compressed, so in the interest of both correctness and safety
+// we should remove all compression markers after we decompress.
+const stripCompressionMarkers = (message: ISequencedDocumentMessage): ISequencedDocumentMessage => {
+    if (message.compression !== undefined) {
+        delete message.compression;
```

Comment:
delete is very expensive in JS. assigning undefined is much less expensive, and can be done unconditionally